### PR TITLE
QStash PoC

### PR DESCRIPTION
**Pull request by ctrlaltdylan**
Sample minimal interactions to interact with QStash to publish messages as callbacks and topics.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4296"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Review by andrewjschuang — APPROVED** [review]
Hi @ctrlaltdylan, I've made some changes so you can quickly continue with your PoC!

Not sure if QA is needed, but I moved this to next the column.

**Comment by DilanAthukorala** [comment]
Hi @ctrlaltdylan The de-duplication id field is expecting a boolean value but the description is a bit misleading. 
image
image

Also the Cron did not executed at the given time.  example cron value: 15 19 * * *

**Comment by ctrlaltdylan** [comment]
Hi @andrewjschuang thanks!

Just a few things I needed to correct:

1. `const headers = {}` needed to be swapped with a `let headers = {}` because we're modifying the object after the variable is created.
2. There are actually two separate headers for controlling content deduplication behavior, I fixed the redundant header
3. I renamed the 2nd header to be suffixed with `Enabled` so it's more clear this is a boolean variable, not a string like the other content deduplication ID variable.

**Review by andrewjschuang — APPROVED** [review]
Thanks, LGTM!

**Comment by DilanAthukorala** [comment]
Verification completed, this is ready for release.

**Comment by ctrlaltdylan** [comment]
Sorry guys, I got carried away and added some more actions + sources.

### Actions
* List Endpoints
* List Topics
* Create Topic
* Delete Topic
* Verify webhook

### Sources
* Create Topic Webhook

This is my first async option + hooks usage in a source, so please let me know if I missed anything!

File: components/qstash/actions/create-endpoint/create-endpoint.mjs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import qstash from "../../qstash.app.mjs";
+
+export default {
+  name: "Create Endpoint",
+  version: "0.0.1",
+  key: "qstash-create-endpoint",
+  description: "",
```
Comment (andrewjschuang) [inline components/qstash/actions/create-endpoint/create-endpoint.mjs:7]: Missing description

File: components/qstash/actions/create-endpoint/create-endpoint.mjs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    endpointUrl: {
+      propDefinition: [
+        qstash,
+        "endpointUrl",
+      ],
+    },
+  },
+  type: "action",
+  methods: {},
```
Comment (andrewjschuang) [inline components/qstash/actions/create-endpoint/create-endpoint.mjs:30]: Not needed

```suggestion
```

File: components/qstash/actions/create-topic/create-topic.mjs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import qstash from "../../qstash.app.mjs";
+
+export default {
+  name: "Create Topic",
+  version: "0.0.1",
+  key: "qstash-create-topic",
+  description: "",
```
Comment (andrewjschuang) [inline components/qstash/actions/create-topic/create-topic.mjs:7]: Missing description

File: components/qstash/actions/create-topic/create-topic.mjs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    qstash,
+    topicName: {
+      propDefinition: [
+        qstash,
+        "topicName",
+      ],
+    },
+  },
+  type: "action",
+  methods: {},
```
Comment (andrewjschuang) [inline components/qstash/actions/create-topic/create-topic.mjs:18]: Not needed

```suggestion
```

File: components/qstash/actions/list-endpoints/list-endpoints.mjs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+export default {
+  name: "List Endpoints",
+  version: "0.0.1",
+  key: "qstash-list-endpoints",
+  description: "Returns all your existing endpoints.",
+  props: {
+    qstash,
+  },
+  type: "action",
+  methods: {},
```
Comment (andrewjschuang) [inline components/qstash/actions/list-endpoints/list-endpoints.mjs:12]: Not needed

```suggestion
```

File: components/qstash/actions/list-topics/list-topics.mjs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+export default {
+  name: "List Topics",
+  version: "0.0.1",
+  key: "qstash-list-topics",
+  description: "Returns all your existing topics.",
+  props: {
+    qstash,
+  },
+  type: "action",
+  methods: {},
```
Comment (andrewjschuang) [inline components/qstash/actions/list-topics/list-topics.mjs:12]: Not needed

```suggestion
```

File: components/qstash/common.mjs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+// import { Receiver } from "@upstash/qstash";
```
Comment (andrewjschuang) [inline components/qstash/common.mjs:1]:
```suggestion
```

**Review by andrewjschuang — COMMENTED** [review]
Just some minor comments!

**Comment by DilanAthukorala** [comment]
Hi @ctrlaltdylan, In Create Endpoint action Type missing for "topic id" prop

image

**Comment by ctrlaltdylan** [comment]
Hi @DilanAthukorala great catch!

I believe the problem was I renamed the `topicId` in the app file, but forgot to replace it. I just readded it. Can you retry?

@andrewjschuang I've just removed the empty `methods` and added the missing descriptions. Thanks!

**Comment by DilanAthukorala** [comment]
> Hi @DilanAthukorala great catch!
> 
> I believe the problem was I renamed the `topicId` in the app file, but forgot to replace it. I just readded it. Can you retry?
> 
> @andrewjschuang I've just removed the empty `methods` and added the missing descriptions. Thanks!

@ctrlaltdylan I'm getting the following error 
image

**Comment by ctrlaltdylan** [comment]
Hi @DilanAthukorala 

I replaced the `topicId` with an easy to use dropdown list with async options instead.

I removed the `topicName` field too.

I tested on my PD account and it's working from what I can see.

**Pull request merged**